### PR TITLE
[Central group table] Split Mapper into two, more specialized classes

### PR DESCRIPTION
**Pull request by mrow4a**
The idea is to reduce very specialized class - Mapper - into two, smaller classes. This should make it is easier to understand and maintain. 

I will also need to have Access class for https://github.com/owncloud/core/pull/29107. The problem there is, that MembershipManager is not really a mapper, but yet still needs some of the Mapper class function. Thus split will be beneficial for that also.

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
Hmmm .... would class Access be of any value on it's own?

**Comment by mrow4a** [comment]
Please look in the main description.

**Comment by mrow4a** [comment]
@jvillafanez @DeepDiver1975 can you have a look now?

File: lib/public/AppFramework/Db/Access.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	 * Runs an sql query
+	 *
+	 * @param string $sql the prepare string
+	 * @param array $params the params which should replace the ? in the sql query
+	 * @param int $limit the maximum number of rows
+	 * @param int $offset from which row we want to start
+	 * @return \PDOStatement the database query result
+	 * @since 10.0.0
+	 */
+	protected function execute($sql, array $params=[], $limit=null, $offset=null) {
```
Comment (mrow4a) [inline lib/public/AppFramework/Db/Access.php]: Access class abstracts from inheriter a way to execute sql query

File: lib/public/AppFramework/Db/Mapper.php [hunk 2]
```diff
@@ -121,7 +123,7 @@ public function insert(Entity $entity){
 		}
 
 		$sql = 'INSERT INTO `' . $this->tableName . '`(' .
-				$columns . ') VALUES(' . $values . ')';
+			$columns . ') VALUES(' . $values . ')';
 
 		$stmt = $this->execute($sql, $params);
```
Comment (mrow4a) [inline lib/public/AppFramework/Db/Mapper.php]: Mapper constructs the sql query, and uses Access to physically execute it.

**Comment by jvillafanez** [comment]
While I understand the purpose of the PR (separate the responsabilities: use "Access" for the connection, and the mapper for the ORM-like stuff), I don't think inheritance is the way to go. Why not inject the "Access" class in the constructor and use it as dependency?

We also need to recheck what is the responsability of the "Access" class in case some other class has already that responsability. What does "Access" do differently compared with "IDBConnection"? If it's just to simplify the calls, could the "Mapper" class absorb that responsability?

**Comment by mrow4a** [comment]
The core thing of Access class is `protected function execute`. Any class, which wants to correctly execute an SQL query, with parameters, limit and offset should inherit from this class. No injection needed.

**Comment by jvillafanez** [comment]
Quoting from https://en.wikipedia.org/wiki/Composition_over_inheritance
> To favor composition over inheritance is a design principle that gives the design higher flexibility. It is more natural to build business-domain classes out of various components than trying to find commonality between them and creating a family tree. For example, a gas pedal and a wheel share very few common traits, yet are both vital components in a car. What they can do and how they can be used to benefit the car is easily defined. Composition also provides a more stable business domain in the long term as it is less prone to the quirks of the family members. In other words, it is better to compose what an object can do (HAS-A) than extend what it is (IS-A).[1]

Unless there is a clear reason to use inheritance, I'd prefer to inject the class instead of inherit from it. Reusing code isn't a valid option to choose inheritance when we have better alternatives

**Comment by mrow4a** [comment]
Ok, @jvillafanez I can see one problem, unfortunately, it is public interface @DeepDiver1975, and other classes are using $db db connection directly from there. I think we need to use inheritance, since injection will break public interface, which unfortunately class Mapper is (we cannot change constructor)

**Comment by mrow4a** [comment]
Ok, I found a better way. I just need to use entirely query builder, which allows me to use limit and offset

**Pull request closed without merging**